File: solutions/2026-09-19/javascript/2026-09-19-circle-and-rectangle-overlapping.js
// Summary: Checks if a circle and an axis-aligned rectangle overlap.
// Link: https://leetcode.com/problems/circle-and-rectangle-overlapping/
// Approach: The core idea is to find the point on the rectangle that is closest to the circle's center.
// If the distance between this closest point and the circle's center is less than or equal to the radius,
// then there is an overlap.
// To find the closest point on the rectangle to the circle's center (xCenter, yCenter):
// 1. For the x-coordinate:
//    - If xCenter is less than x1, the closest x on the rectangle is x1.
//    - If xCenter is greater than x2, the closest x on the rectangle is x2.
//    - Otherwise (xCenter is between x1 and x2), the closest x on the rectangle is xCenter itself.
// 2. For the y-coordinate:
//    - If yCenter is less than y1, the closest y on the rectangle is y1.
//    - If yCenter is greater than y2, the closest y on the rectangle is y2.
//    - Otherwise (yCenter is between y1 and y2), the closest y on the rectangle is yCenter itself.
// Once the closest point (closestX, closestY) is found, calculate the squared distance between
// (xCenter, yCenter) and (closestX, closestY): `distX*distX + distY*distY`.
// Compare this squared distance with the squared radius (`radius*radius`). If it's less than or equal to, they overlap.
// Using squared distances avoids the costly `Math.sqrt()` operation.
// Time Complexity: O(1) - The calculations involve a fixed number of operations regardless of input size.
// Space Complexity: O(1) - The algorithm uses a constant amount of extra space for variables.

/**
 * @param {number} radius
 * @param {number} xCenter
 * @param {number} yCenter
 * @param {number} x1
 * @param {number} y1
 * @param {number} x2
 * @param {number} y2
 * @return {boolean}
 */
var checkOverlap = function(radius, xCenter, yCenter, x1, y1, x2, y2) {
    // Find the closest x-coordinate on the rectangle to the circle's center.
    // If xCenter is to the left of the rectangle, closestX is x1.
    // If xCenter is to the right of the rectangle, closestX is x2.
    // Otherwise, xCenter is within the rectangle's x-bounds, so closestX is xCenter.
    const closestX = Math.max(x1, Math.min(xCenter, x2));

    // Find the closest y-coordinate on the rectangle to the circle's center.
    // If yCenter is below the rectangle, closestY is y1.
    // If yCenter is above the rectangle, closestY is y2.
    // Otherwise, yCenter is within the rectangle's y-bounds, so closestY is yCenter.
    const closestY = Math.max(y1, Math.min(yCenter, y2));

    // Calculate the distance between the circle's center and the closest point on the rectangle.
    // We are using squared distances to avoid the `Math.sqrt()` operation for efficiency.
    const distX = xCenter - closestX;
    const distY = yCenter - closestY;

    // Calculate the squared distance.
    const distanceSquared = (distX * distX) + (distY * distY);

    // The circle and rectangle overlap if the squared distance between the circle's center
    // and the closest point on the rectangle is less than or equal to the squared radius.
    return distanceSquared <= (radius * radius);
};

File: solutions/2026-09-19/cpp/2026-09-19-circle-and-rectangle-overlapping.cpp
```cpp
// Checks if a circle and an axis-aligned rectangle overlap.
// Link: https://leetcode.com/problems/circle-and-rectangle-overlapping/
// Approach: The problem boils down to finding the closest point on the rectangle to the circle's center.
// If the distance between the circle's center and this closest point is less than or equal to the circle's radius,
// then there is an overlap.
// To find the closest point on the rectangle to the circle's center (xCenter, yCenter):
// 1. Clamp the circle's x-coordinate to the rectangle's x-range [x1, x2]. Let this be closestX.
//    If xCenter < x1, closestX = x1.
//    If xCenter > x2, closestX = x2.
//    Otherwise, closestX = xCenter.
// 2. Clamp the circle's y-coordinate to the rectangle's y-range [y1, y2]. Let this be closestY.
//    If yCenter < y1, closestY = y1.
//    If yCenter > y2, closestY = y2.
//    Otherwise, closestY = yCenter.
// The closest point on the rectangle to the circle's center is (closestX, closestY).
// Then, calculate the distance between (xCenter, yCenter) and (closestX, closestY) using the distance formula:
// distance = sqrt((xCenter - closestX)^2 + (yCenter - closestY)^2)
// If distance <= radius, return true. Otherwise, return false.
// Time Complexity: O(1) - The calculations involve a fixed number of operations, independent of input size.
// Space Complexity: O(1) - Only a few variables are used to store intermediate results.
class Solution {
public:
    bool checkOverlap(int radius, int xCenter, int yCenter, int x1, int y1, int x2, int y2) {
        // Find the closest point to the circle's center that lies on the rectangle.

        // Clamp the circle's x-coordinate to be within the rectangle's x-range [x1, x2].
        // If xCenter is less than x1, the closest x-coordinate on the rectangle is x1.
        // If xCenter is greater than x2, the closest x-coordinate on the rectangle is x2.
        // Otherwise, the closest x-coordinate is xCenter itself (meaning the circle's center's x is within the rectangle's x bounds).
        int closestX = std::max(x1, std::min(xCenter, x2));

        // Clamp the circle's y-coordinate to be within the rectangle's y-range [y1, y2].
        // Similar logic as for the x-coordinate.
        int closestY = std::max(y1, std::min(yCenter, y2));

        // Calculate the distance between the circle's center (xCenter, yCenter) and the closest point on the rectangle (closestX, closestY).
        // We use the squared distance to avoid using `sqrt` which can be computationally more expensive and potentially introduce floating-point inaccuracies.
        // The condition distance <= radius is equivalent to distance^2 <= radius^2.
        long long distanceX = (long long)xCenter - closestX;
        long long distanceY = (long long)yCenter - closestY;
        long long squaredDistance = (distanceX * distanceX) + (distanceY * distanceY);

        // If the squared distance is less than or equal to the squared radius, the circle and rectangle overlap.
        return squaredDistance <= (long long)radius * radius;
    }
};
```